File: samples/java/ftl/TibMonSub.java
/*
 * Copyright (c) 2010-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample of a basic java FTL monitoring subscriber program.
 */

package com.tibco.ftl.samples;

import java.util.*;

import com.tibco.ftl.*;

public class TibMonSub implements SubscriberListener
{
    String   realmServer    = null;
    String   defaultUrl     = "http://localhost:8080";
    String   defaultSecureUrl = "https://localhost:8080";
    String   match          = null;
    String   trcLevel       = null;
    String   user           = "guest";
    String   password       = "guest-pw";
    String   secondary      = null;
    boolean  trustAll       = false;
    String   trustFile      = null;
    String   clientLabel    = null;
    boolean  running        = true;

    String   usageString    =
        "TibMonSub [options] url\n" +
        " Default url is " + defaultUrl + "\n" +
        "  where options can be:\n" +
        "    -cl, --client-label <string>\n" +
        "    -h, --help\n" +
        "    -m, --match <content match>\n" +
        "           where <content match> is of the form:\n" +
        "                 '{\"fieldname\":<value>}'\n" +
        "           where <value> is:\n" +
        "                 \"string\" for a string field\n" +
        "                 a number such as 2147483647 for an integer field\n" +
        "    -p, --password <string>\n" +
        "    -s, --secondary <string>\n" +
        "    -t, --trace <level>\n" +
        "           where <level> can be:\n" +
        "                 off, severe, warn, info, debug, verbose\n" +
        "        --trustall\n" +
        "        --trustfile <file>\n" +
        "    -u, --user <string>\n";

    public TibMonSub(String[] args)
    {
        System.out.printf("#\n# %s\n#\n# %s\n#\n",
                          this.getClass().getName(),
                          FTL.getVersionInformation());

        parseArgs(args);
    }
    
    public void usage()
    {
        System.out.printf(usageString);
        System.exit(0);
    }

    public void parseArgs(String args[])
    {
        int i           = 0;
        int argc        = args.length;
        String s        = null;
        int n           = 0;
        TibAux aux      = new TibAux(args, usageString);

        for (i = 0; i < argc; i++)
        {
            s = aux.getString(i, "--client-label", "-cl");
            if (s != null) {
                clientLabel = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--help", "-h")) {
                usage();
            }
            s = aux.getString(i, "--match", "-m");
            if (s != null) {
                match = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--password", "-p");
            if (s != null) {
                password = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--secondary", "-s");
            if (s != null) {
                secondary = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--trace", "-t");
            if (s != null) {
                trcLevel = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--trustall", "--trustall")) {
                if (trustFile != null)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                trustAll = true;
                continue;
            }
            s = aux.getString(i, "--trustfile", "--trustfile");
            if (s != null)
            {
                trustFile = s;
                if (trustAll)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                i++;
                continue;
            }
            s = aux.getString(i, "--user", "-u");
            if (s != null) {
                user = s;
                i++;
                continue;
            }
            if (args[i].startsWith("-", 0)) {
                System.out.println("invalid option: "+args[i]);
                usage();
            }
            realmServer = args[i];
        }

        System.out.print("Invoked as: TibMonSub ");
        for (i = 0; i < argc; i++)
            System.out.print(args[i]+" ");
        System.out.println("");
    }

    public void processMessageFields(
        Message         msg,
        long            indent) throws FTLException
    {
        boolean addSeparator = false;
        MessageIterator it = msg.iterator();

        while (indent-- > 0)
        {
            System.out.print("    ");
        }
        while (it.hasNext())
        {
            MessageFieldRef ref = it.next();
            int fieldType = (int) msg.getFieldType(ref);
            switch (fieldType)
            {
                case Message.TIB_FIELD_TYPE_LONG_ARRAY:
                case Message.TIB_FIELD_TYPE_DOUBLE_ARRAY:
                case Message.TIB_FIELD_TYPE_STRING_ARRAY:
                case Message.TIB_FIELD_TYPE_MESSAGE_ARRAY:
                case Message.TIB_FIELD_TYPE_DATETIME_ARRAY:
                    continue;
                default:
                    break;
            }
            String fieldName = ref.getFieldName();
            if (addSeparator)
            {
                System.out.print(", ");
            }
            addSeparator = true;
            System.out.print(fieldName);
            System.out.print("=");
            switch (fieldType)
            {
                case Message.TIB_FIELD_TYPE_LONG:
                    System.out.format("%d", msg.getLong(ref));
                    break;
                case Message.TIB_FIELD_TYPE_STRING:
                    System.out.print(msg.getString(ref));
                    break;
                default:
                    System.out.print("Unsupported field type");
                    break;
            }
        }
        System.out.println();
    }

    public void processFTLMetricMessage(
        Message         msg) throws FTLException
    {
        Message[] metrics;

        if (!msg.isFieldSet(Monitoring.FIELD_METRICS))
        {
            return;
        }
        processMessageFields(msg, 0);
        metrics = msg.getMessageArray(Monitoring.FIELD_METRICS);
        for (Message metricMsg : metrics)
        {
            processMessageFields(metricMsg, 1);
        }
    }

    public void processFTLEventMessage(
        Message         msg) throws FTLException
    {
        processMessageFields(msg, 0);
    }

    public void messagesReceived(List<Message> messages, EventQueue eventQueue)
    {
        int i;
        int msgNum = messages.size();
         
        for (i = 0;  i < msgNum;  i++) 
        {
            try
            {
                Message msg = messages.get(i);
                if (msg.isFieldSet(Monitoring.FIELD_MSG_TYPE))
                {
                    long msgType = msg.getLong(Monitoring.FIELD_MSG_TYPE);
                    if (msgType == Monitoring.MessageType.METRICS)
                    {
                        System.out.print("Metrics: ");
                        processFTLMetricMessage(msg);
                    }
                    else if (msgType == Monitoring.MessageType.SERVER_METRICS)
                    {
                        System.out.print("Server metrics: ");
                        processFTLMetricMessage(msg);
                    }
                    else if (msgType == Monitoring.MessageType.EVENT)
                    {
                        System.out.print("Event: ");
                        processFTLEventMessage(msg);
                    }
                }
            }
            catch(FTLException exp)
            {
                running = false;
                exp.printStackTrace();
            }
        }
    }
    
    public int recv() throws FTLException
    {
        TibProperties   props;
        Realm           realm;
        ContentMatcher  cm = null;
        ContentMatcher  am;
        Subscriber      sub;
        Subscriber      adv;
        EventQueue      queue;
        
        // Set global trace to specified level
        if (trcLevel != null)
            FTL.setLogLevel(trcLevel);
        
        // Get a single realm per process
        props = FTL.createProperties();
        props.set(Realm.PROPERTY_STRING_USERNAME, user);
        props.set(Realm.PROPERTY_STRING_USERPASSWORD, password);
        if (secondary != null)
            props.set(Realm.PROPERTY_STRING_SECONDARY_SERVER, secondary);

        if (trustAll || trustFile != null)
        {
            if (realmServer == null)
                realmServer = defaultSecureUrl;

            if (trustAll)
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_TRUST_EVERYONE);
            }
            else
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
                props.set(Realm.PROPERTY_STRING_TRUST_FILE, trustFile);
            }
        }
        else
        {
            if (realmServer == null)
                realmServer = defaultUrl;
        }

        if (clientLabel != null)
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, clientLabel);
        else
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, this.getClass().getSimpleName());

        realm = FTL.connectToRealmServer(realmServer, null, props);
        props.destroy();
        
        // Create a content matcher, if specified
        if (match != null)
            cm = realm.createContentMatcher(match);

        // Create a subscriber
        sub = realm.createSubscriber(Monitoring.ENDPOINT_NAME, cm, null);

        // Create a data loss advisory matcher
        am = realm.createContentMatcher("{\""+Advisory.FIELD_NAME+
                                        "\":\""+Advisory.NAME_DATALOSS+"\"}");

        // Create an advisory subscriber
        adv = realm.createSubscriber(Advisory.ENDPOINT_NAME, am);
        am.destroy();

        // Create a queue and add subscriber to it
        queue = realm.createEventQueue();
        queue.addSubscriber(sub, this);
        queue.addSubscriber(adv, new AdvisoryListener());
        
        // Begin dispatching messages
        System.out.printf("waiting for message(s)\n");
        while (running)
        {
            queue.dispatch();
        }

        // Clean up
        queue.removeSubscriber(sub);
        queue.removeSubscriber(adv);
        queue.destroy();
        sub.close();
        adv.close();
        if (cm != null)
            cm.destroy();
        realm.close();

        return 0;
    }
    
    public static void main(String[] args)
    {
        TibMonSub s  = new TibMonSub(args);
        try
        {
            s.recv();
        }
        catch (Throwable e)
        {
            e.printStackTrace();
        }
    } 

    private class AdvisoryListener implements SubscriberListener
    {
        public void messagesReceived(List<Message> messages, EventQueue eventQueue)
        {
            int msgNum = messages.size();

            for (int i = 0;  i < msgNum;  i++)
            {
                Message msg = messages.get(i);
                try
                {
                    System.out.println("advisory:");

                    System.out.println("Name: " + msg.getString(Advisory.FIELD_NAME));

                    System.out.println("    Severity: " +
                                       msg.getString(Advisory.FIELD_SEVERITY));
                    System.out.println("    Module: " +
                                       msg.getString(Advisory.FIELD_MODULE));
                    System.out.println("    Reason: " +
                                       msg.getString(Advisory.FIELD_REASON));
                    System.out.println("    Timestamp: " +
                                       msg.getDateTime(Advisory.FIELD_TIMESTAMP));
                    System.out.println("    Aggregation Count: " +
                                       msg.getLong(Advisory.FIELD_AGGREGATION_COUNT));
                    System.out.println("    Aggregation Time: " +
                                       msg.getDouble(Advisory.FIELD_AGGREGATION_TIME));

                    if (msg.isFieldSet(Advisory.FIELD_ENDPOINTS))
                    {
                        System.out.println("    Endpoints:");
                        String[] sarray = msg.getStringArray(Advisory.FIELD_ENDPOINTS);
                        for (int j = 0; j < sarray.length; j++)
                            System.out.println("      " + sarray[j]);
                    }

                    if (msg.isFieldSet(Advisory.FIELD_SUBSCRIBER_NAME))
                    {
                        System.out.println("  Subscriber Name: " +
                                           msg.getString(Advisory.FIELD_SUBSCRIBER_NAME));
                    }
                }
                catch(FTLException exp)
                {
                    exp.printStackTrace();
                }
            }
        }
    }
}

File: samples/c/ftl/tibmonsub.c
/*
 * Copyright (c) 2010-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample of a basic C FTL monitoring subscriber program.
 */

#include <stdlib.h>
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include <ctype.h>

#include "tib/ftl.h"

#include "tibaux.h"

#define DEFAULT_URL "http://localhost:8080"
#define DEFAULT_SECURE_URL "https://localhost:8080"

const char      *appName        = NULL;
const char      *realmServer    = NULL;
const char      *match          = NULL;
const char      *trcLevel       = NULL;
const char      *user           = "guest";
const char      *password       = "guest-pw";
const char      *secondary      = NULL;
tibbool_t       trustAll        = tibfalse;
const char      *trustFile      = NULL;
const char      *clientLabel    = NULL;

tibFieldRef     refMsgType             = NULL;
tibFieldRef     refMetrics             = NULL;

void
usage(
    int         argc,
    char        **argv)
{
    printf("Usage: %s [options] url\n", argv[0]);
    printf("Default url is "DEFAULT_URL"\n");
    printf("    -a,  --application <name>\n"
           "    -d,  --durablename <name>\n"
           "         --client-label <label>\n"
           "    -h,  --help\n"
           "    -m,  --match <content match>\n"
           "            where <content match> is of the form:\n"
           "                 '{\"fieldname\":<value>}'\n"
           "            where <value> is:\n"
           "                 \"string\" for a string field\n"
           "                 a number such as 2147483647 for an integer field\n"
           "    -p,  --password <string>\n"
           "    -u,  --user <string>\n"
           "    -s,  --secondary <string>\n"
           "    -t,  --trace <level>\n"
           "            where <level> can be:\n"
           "                  off, severe, warn, info, debug, verbose\n"
           "         --trustall\n"
           "         --trustfile <file>\n"
           );

    exit(1);
}

void
parseArgs(int argc, char** argv)
{
    int    i = 1;

    tibAux_InitGetArgs(argc, argv, usage);

    for (i = 1;  i < argc;  i++)
    {
        if (tibAux_GetString(&i, "--application", "-a", &appName))
            continue;
        if (tibAux_GetString(&i, "--client-label", "--client-label", &clientLabel))
            continue;
        if (tibAux_GetFlag(&i, "--help", "-h"))
            usage(argc, argv);
        if (tibAux_GetString(&i, "--match", "-m", &match))
            continue;
        if (tibAux_GetString(&i, "--password", "-p", &password))
            continue;
        if (tibAux_GetString(&i, "--secondary", "-s", &secondary))
            continue;
        if (tibAux_GetString(&i, "--trace", "-t", &trcLevel))
            continue;
        if (tibAux_GetString(&i, "--user", "-u", &user))
            continue;
        if (tibAux_GetFlag(&i, "--trustall", "--trustall")) {
            if (trustFile)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            trustAll = tibtrue;
            continue;
        }
        if (tibAux_GetString(&i, "--trustfile", "--trustfile", &trustFile))
        {
            if (trustAll)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            continue;
        }
        if (strncmp(argv[i], "-", 1)==0)
        {
            printf("invalid option: %s\n", argv[i]);
            usage(argc, argv);
        }
        realmServer = argv[i];
    }

    printf("Invoked as: ");
    for (i = 0;  i < argc;  i++)
        printf("%s ", argv[i]);
    printf("\n");
}

void
memFail(
    const char          *obj)
{
    printf("memory allocation failed for %s\n", obj);
    exit(-1);
}

void
createMsgFieldRefs(
    tibEx               ex)
{
    refMsgType                 = tibFieldRef_Create(ex, TIB_MONITORING_FIELD_MSG_TYPE);
    refMetrics                 = tibFieldRef_Create(ex, TIB_MONITORING_FIELD_METRICS);
}

void
cleanupMsgFieldRefs(
    tibEx               ex)
{
    tibFieldRef_Destroy(ex, refMsgType);
    tibFieldRef_Destroy(ex, refMetrics);
}

void
processMessageFields(
    tibEx               ex,
    tibMessage          msg,
    unsigned int        indent)
{
    tibbool_t           addSeparator = tibfalse;
    tibMessageIterator  it = NULL;

    while (indent-- > 0)
    {
        fputs("    ", stdout);
    }
    it = tibMessageIterator_Create(ex, msg);
    while (tibMessageIterator_HasNext(ex, it))
    {
        tibFieldRef ref = NULL;
        const char * fieldName = NULL;
        tibFieldType fieldType;
        const char * stringVal;
        tibint64_t intVal;
        char valBuf[32];

        ref = tibMessageIterator_GetNext(ex, it);
        fieldType = tibMessage_GetFieldTypeByRef(ex, msg, ref);
        switch (fieldType)
        {
            case TIB_FIELD_TYPE_LONG_ARRAY:
            case TIB_FIELD_TYPE_DOUBLE_ARRAY:
            case TIB_FIELD_TYPE_STRING_ARRAY:
            case TIB_FIELD_TYPE_MESSAGE_ARRAY:
            case TIB_FIELD_TYPE_DATETIME_ARRAY:
                /* Skip array fields */
                continue;
            default:
                break;
        }
        fieldName = tibFieldRef_GetFieldName(ex, ref);
        if (addSeparator)
        {
            fputs(", ", stdout);
        }
        addSeparator = tibtrue;
        fputs(fieldName, stdout);
        fputs("=", stdout);
        switch (fieldType)
        {
            case TIB_FIELD_TYPE_LONG:
                intVal = tibMessage_GetLongByRef(ex, msg, ref);
                tibAux_snprintf(valBuf, sizeof(valBuf), "%" auxPRId64, intVal);
                fputs(valBuf, stdout);
                break;
            case TIB_FIELD_TYPE_STRING:
                stringVal = tibMessage_GetStringByRef(ex, msg, ref);
                fputs(stringVal, stdout);
                break;
            default:
                fputs("Unsupported field type", stdout);
                break;
        }
    }
    tibMessageIterator_Destroy(ex, it);
    fputs("\n", stdout);
}

void
processFTLMetricMessage(
    tibEx               ex,
    tibMessage          msg)
{
    tibMessage*         metrics = NULL;
    tibint32_t          metricsCount;
    tibint32_t          idx;

    if (!tibMessage_IsFieldSetByRef(ex, msg, refMetrics))
    {
        return;
    }
    processMessageFields(ex, msg, 0);
    metrics = (tibMessage *) tibMessage_GetArrayByRef(ex, msg, TIB_FIELD_TYPE_MESSAGE_ARRAY, refMetrics, &metricsCount);
    for (idx = 0; idx < metricsCount; ++idx)
    {
        processMessageFields(ex, metrics[idx], 1);
    }
}

void
processFTLEventMessage(
    tibEx               ex,
    tibMessage          msg)
{
    processMessageFields(ex, msg, 0);
}

void
onMessages(
    tibEx               ex,
    tibEventQueue       msgQueue,
    tibint32_t          msgNum,
    tibMessage          *msgs,
    void                **closures)
{
    tibint32_t          i;

    for (i = 0; i < msgNum; i++)
    {
        tibMessage msg = msgs[i];
        if (tibMessage_IsFieldSetByRef(ex, msg, refMsgType))
        {
            tibint64_t msgType = tibMessage_GetLongByRef(ex, msg, refMsgType);
            switch (msgType)
            {
                case TIB_MONITORING_MSG_TYPE_METRICS:
                    fputs("Metrics: ", stdout);
                    processFTLMetricMessage(ex, msg);
                    break;
                case TIB_MONITORING_MSG_TYPE_SERVER_METRICS:
                    fputs("Server metrics: ", stdout);
                    processFTLMetricMessage(ex, msg);
                    break;
                case TIB_MONITORING_MSG_TYPE_EVENT:
                    fputs("Event: ", stdout);
                    processFTLEventMessage(ex, msg);
                    break;
                default:
                    break;
            }
        }
    }

    // any exception that happens in the callback does not
    // propagate to the caller of the tibEventQueue_Dispatch
    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
}

void
onAdvisory(
    tibEx               ex,
    tibEventQueue       msgQueue,
    tibint32_t          msgNum,
    tibMessage          *msgs,
    void                **closures)
{
    tibint32_t          i;

    for (i = 0; i < msgNum; i++)
    {
        printf("received advisory\n");
        tibAux_PrintAdvisory(ex, msgs[i]);
    }

    // any exception that happens in the callback does not
    // propagate to the caller of the tibEventQueue_Dispatch
    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
}

int
main(int argc, char** argv)
{
    tibEx               ex;
    tibProperties       props;
    tibRealm            realm;
    tibEventQueue       queue = NULL;
    tibContentMatcher   cm = NULL;
    tibContentMatcher   am;
    tibSubscriber       sub;
    tibSubscriber       adv;
    char                str[256];

    printf("#\n# %s\n#\n# %s\n#\n", argv[0], tib_Version());

    parseArgs(argc, argv);

    ex = tibEx_Create();
    if (ex == NULL)
        memFail("ex");

    tib_Open(ex, TIB_COMPATIBILITY_VERSION);
    CHECK(ex);

    // Set global trace to specified level
    if (trcLevel != NULL)
        tib_SetLogLevel(ex, trcLevel);

    // Get a single realm per process
    props = tibProperties_Create(ex);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERNAME, user);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERPASSWORD, password);
    if (secondary)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_SECONDARY_SERVER, secondary);

    if (trustAll || trustFile)
    {
        if (!realmServer)
            realmServer = DEFAULT_SECURE_URL;

        if (trustAll)
        {
            tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                                        TIB_REALM_HTTPS_CONNECTION_TRUST_EVERYONE);
        }
        else
        {
            tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                    TIB_REALM_HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
            tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_TRUST_FILE, trustFile);
        }
    }
    else
    {
        if (!realmServer)
            realmServer = DEFAULT_URL;
    }

    // set the client label
    if (clientLabel != NULL)
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, clientLabel);
    else
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, argv[0]);

    realm = tibRealm_Connect(ex, realmServer, appName, props);
    CHECK(ex);
    tibProperties_Destroy(ex, props);

    // Create a content matcher, if specified
    if (match != NULL)
        cm = tibContentMatcher_Create(ex, realm, match);

    // Create a subscriber
    sub = tibSubscriber_Create(ex, realm, TIB_MONITORING_ENDPOINT_NAME, cm, NULL);
    CHECK(ex);

    // Create a data loss advisory matcher 
    tibAux_snprintf(str, sizeof(str), "{\"%s\":\"%s\"}", 
                  TIB_ADVISORY_FIELD_NAME, TIB_ADVISORY_NAME_DATALOSS);
    am = tibContentMatcher_Create(ex, realm, str);

    // Create an advisory subscriber
    adv = tibSubscriber_Create(ex, realm, TIB_ADVISORY_ENDPOINT_NAME, am, NULL);
    CHECK(ex);

    // Create a queue and add subscribers to it
    queue = tibEventQueue_Create(ex, realm, NULL);
    tibEventQueue_AddSubscriber(ex, queue, sub, onMessages, NULL);
    tibEventQueue_AddSubscriber(ex, queue, adv, onAdvisory, NULL);
    CHECK(ex);

    // set up references to the expected message format for fast access
    createMsgFieldRefs(ex);

    // Begin dispatching messages
    printf("waiting for message(s)\n");
    while (1)
    {
        tibEventQueue_Dispatch(ex, queue, TIB_TIMEOUT_WAIT_FOREVER);
        if (tibEx_GetErrorCode(ex) != TIB_OK)
            break;
    }

    // Clean up
    cleanupMsgFieldRefs(ex);
    if (queue && sub)
        tibEventQueue_RemoveSubscriber(ex, queue, sub, NULL);
    if (queue && adv)
        tibEventQueue_RemoveSubscriber(ex, queue, adv, NULL);
    tibEventQueue_Destroy(ex, queue, NULL);
    tibSubscriber_Close(ex, sub);
    tibSubscriber_Close(ex, adv);
    tibContentMatcher_Destroy(ex, cm);
    tibContentMatcher_Destroy(ex, am);

    tibRealm_Close(ex, realm);
    tib_Close(ex);

    // Check if it worked before destroying the exception
    CHECK(ex);
    tibEx_Destroy(ex);

    return 0;
}
